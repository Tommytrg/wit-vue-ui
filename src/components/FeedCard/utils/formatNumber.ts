export function formatNumber(num: string): string {
  const parsedNum = parseFloat(num).toString()
  const splitedNumber = parsedNum.split('.')
  parseFloat(`0.${splitedNumber[1]}`)
  const decimals = splitedNumber.length > 1 ? '.' + splitedNumber[1] : ''
  const rgx = /(\d)(?=(\d{3})+(?!\d))/g
  const unit = splitedNumber[0].replace(rgx, '$1,')
  return unit + decimals
}
