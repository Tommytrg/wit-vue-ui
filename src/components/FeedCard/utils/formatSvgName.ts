export function formatSvgName(name: string): string {
  return name.split('/')[0]
}
