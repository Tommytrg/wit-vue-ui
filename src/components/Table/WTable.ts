export enum Sort {
  ascentant,
  alphabetically,
  alphabeticallyReverse,
  descendant
}

export type Column = {
  value: string | number
  label: string
  url: string | null
}

export type Label = {
  sortType: Sort
  label: string
  index: number
}

export type Row = Column[]
