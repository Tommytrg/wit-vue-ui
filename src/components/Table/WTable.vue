<template>
  <table
    class="bg-white-100 dark:bg-black-800 rounded-[24px] w-full divide-y sm:divide-y-0 divide-white-200 dark:divide-white-500 text-black-500 mb-md sm:w-full"
  >
    <thead class="text-xs dark:text-white-200 sm:hidden text-black-500">
      <tr>
        <th
          scope="col"
          class="px-md py-md text-start cursor-pointer last:text-end"
          v-for="col in labels"
          :key="col.label"
          @click="setSortLabel(col.sortType, col.index)"
        >
          {{ col.label }}
          <WSortArrow
            :current-label="currentSortType"
            :label="col.sortType"
            :order="order"
          ></WSortArrow>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        v-for="row in sortedData"
        :key="row[0].value"
        class="transition-all duration-200 even:bg-white-200 dark:even:bg-black-600 sm:grid sm:grid-col-1 sm:py-md"
      >
        <div v-for="col in row" :key="col.value" class="responsive-table">
          <th class="label">{{ col.label }}</th>
          <td class="px-md py-md [&&]:sm:pt-sm whitespace-nowrap">
            <WColumn :url="col.url ?? undefined" :value="col.value"></WColumn>
          </td>
        </div>
        <td
          v-for="col in row"
          :key="col.value"
          class="sm:hidden px-md py-md [&&]:sm:pt-sm whitespace-nowrap last:text-end"
        >
          <WColumn :url="col.url ?? undefined" :value="col.value"></WColumn>
        </td>
      </tr>
    </tbody>
  </table>
  <div v-if="loading" class="text-center py-4">
    <div class="inline-block animate-spin rounded-full h-8 w-8 border-b-2 border-wit-blue-500" />
  </div>
</template>

<script setup lang="ts">
import { ref, computed, type Ref } from 'vue'
import { Label, Row, Sort } from './WTable'
import WSortArrow from './WSortArrow.vue'
import WColumn from './WColumn.vue'

const props = defineProps({
  loading: Boolean,
  labels: {
    type: Array<Label>,
    required: true
  },
  data: {
    type: Array<Row>,
    required: true
  }
})

const currentSortType: Ref<Sort> = ref(Sort.ascentant)
const indexToSort: Ref<number> = ref(0)
const order: Ref<boolean> = ref(true)

const sortedData: Ref<Row[]> = computed(() => {
  const data: Row[] = props.data
  console.log(indexToSort)
  const sortFunctions = {
    [Sort.ascentant]: (a: Row, b: Row) => {
      const aValue: number = a[indexToSort.value].value as number
      const bValue: number = b[indexToSort.value].value as number
      return order.value ? aValue - bValue : bValue - aValue
    },
    [Sort.descendant]: (a: Row, b: Row) => {
      const aValue: number = a[indexToSort.value].value as number
      const bValue: number = b[indexToSort.value].value as number
      return order.value ? bValue - aValue : aValue - bValue
    },
    [Sort.alphabetically]: (a: Row, b: Row) => {
      const aValue: string = a[indexToSort.value].value as string
      const bValue: string = b[indexToSort.value].value as string
      return order.value ? aValue.localeCompare(bValue) : bValue.localeCompare(aValue)
    },
    [Sort.alphabeticallyReverse]: (a: Row, b: Row) => {
      const aValue: string = a[indexToSort.value].value as string
      const bValue: string = b[indexToSort.value].value as string
      return order.value ? bValue.localeCompare(aValue) : aValue.localeCompare(bValue)
    }
  }
  return data.sort(sortFunctions[currentSortType.value])
})

function setSortLabel(label: Sort, index: number) {
  if (currentSortType.value === label) {
    order.value = !order.value
    indexToSort.value = index
  } else {
    currentSortType.value = label
    order.value = true
    indexToSort.value = index
  }
}
</script>
<style lang="scss" scoped>
.pointer-events-none {
  @apply text-black-950;
}
.responsive-table {
  @apply hidden sm:grid;
}
.label {
  @apply px-md font-bold text-sm text-black-400 text-start;
}
.t-value {
  @apply px-md py-md [&&]:sm:pt-sm;
}
</style>
