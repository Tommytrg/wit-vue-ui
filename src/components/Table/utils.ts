import { Row, Sort } from './WTable'

export function getMockupLabels() {
  return [
    {
      sortType: Sort.ascentant,
      label: 'Rank',
      index: 0
    },
    {
      sortType: Sort.alphabetically,
      label: 'Address',
      index: 1
    },
    {
      sortType: Sort.descendant,
      label: 'Amount',
      index: 2
    }
  ]
}
const rankRow = {
  value: 1,
  label: 'Rank',
  url: null
}
const addressRow = {
  value: 'wit3',
  label: 'Address',
  url: 'https://'
}
const amountRow = {
  value: 1,
  label: 'Amount',
  url: null
}
const row: Row = [rankRow, addressRow, amountRow]
export function addRow(index: number): Row {
  return row.map((col) => {
    if (typeof col.value == 'number') {
      return {
        ...col,
        value: index
      }
    } else {
      return {
        ...col,
        value: `wit${index - 1}`
      }
    }
  })
}
export function getMockupData() {
  const row2: Row = addRow(2)
  const row3: Row = addRow(3)
  return [row, row2, row3]
}
