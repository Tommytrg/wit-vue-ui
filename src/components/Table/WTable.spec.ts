import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import WTable from './WTable.vue'
import { getMockupData, getMockupLabels } from './utils'

describe('WTable', () => {
  it('renders properly', () => {
    const wrapper = mount(WTable, {
      props: {
        labels: getMockupLabels(),
        data: getMockupData()
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })
})
