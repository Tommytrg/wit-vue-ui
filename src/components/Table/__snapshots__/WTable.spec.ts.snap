// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WTable > renders properly 1`] = `
<div
  data-v-app=""
>
  
  <table
    class="bg-white-100 dark:bg-black-800 rounded-[24px] w-full divide-y sm:divide-y-0 divide-white-200 dark:divide-white-500 text-black-500 mb-md sm:w-full"
    data-v-a63dd56f=""
  >
    <thead
      class="text-xs dark:text-white-200 sm:hidden text-black-500"
      data-v-a63dd56f=""
    >
      <tr
        data-v-a63dd56f=""
      >
        
        <th
          class="px-md py-md text-start cursor-pointer last:text-end"
          data-v-a63dd56f=""
          scope="col"
        >
          Rank 
          <svg
            class="color w-[10px] h-auto pl-xs inline w-[10px] h-auto pl-xs inline"
            data-v-a63dd56f=""
            data-v-c15836e8=""
            fill="none"
            height="11"
            viewBox="0 0 11 11"
            width="11"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.799 9.75a1.5 1.5 0 0 1-2.598 0L.304 3A1.5 1.5 0 0 1 1.603.75h7.794a1.5 1.5 0 0 1 1.3 2.25z"
              fill="#D9D9D9"
            />
          </svg>
        </th>
        <th
          class="px-md py-md text-start cursor-pointer last:text-end"
          data-v-a63dd56f=""
          scope="col"
        >
          Address 
          <svg
            class="color w-[10px] h-auto pl-xs inline w-[10px] h-auto pl-xs inline"
            data-v-a63dd56f=""
            data-v-c15836e8=""
            fill="none"
            height="11"
            viewBox="0 0 11 11"
            width="11"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.799 9.75a1.5 1.5 0 0 1-2.598 0L.304 3A1.5 1.5 0 0 1 1.603.75h7.794a1.5 1.5 0 0 1 1.3 2.25z"
              fill="#D9D9D9"
            />
          </svg>
        </th>
        <th
          class="px-md py-md text-start cursor-pointer last:text-end"
          data-v-a63dd56f=""
          scope="col"
        >
          Amount 
          <svg
            class="color w-[10px] h-auto pl-xs inline w-[10px] h-auto pl-xs inline"
            data-v-a63dd56f=""
            data-v-c15836e8=""
            fill="none"
            height="11"
            viewBox="0 0 11 11"
            width="11"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.799 9.75a1.5 1.5 0 0 1-2.598 0L.304 3A1.5 1.5 0 0 1 1.603.75h7.794a1.5 1.5 0 0 1 1.3 2.25z"
              fill="#D9D9D9"
            />
          </svg>
        </th>
        
      </tr>
    </thead>
    <tbody
      data-v-a63dd56f=""
    >
      
      <tr
        class="transition-all duration-200 even:bg-white-200 dark:even:bg-black-600 sm:grid sm:grid-col-1 sm:py-md"
        data-v-a63dd56f=""
      >
        
        <div
          class="responsive-table"
          data-v-a63dd56f=""
        >
          <th
            class="label"
            data-v-a63dd56f=""
          >
            Rank
          </th>
          <td
            class="px-md py-md [&&]:sm:pt-sm whitespace-nowrap"
            data-v-a63dd56f=""
          >
            <span
              data-v-a63dd56f=""
            >
              1
            </span>
          </td>
        </div>
        <div
          class="responsive-table"
          data-v-a63dd56f=""
        >
          <th
            class="label"
            data-v-a63dd56f=""
          >
            Address
          </th>
          <td
            class="px-md py-md [&&]:sm:pt-sm whitespace-nowrap"
            data-v-a63dd56f=""
          >
            <a
              class="text-wit-blue-600 dark:text-wit-blue-500"
              data-v-a63dd56f=""
              href="https://"
              target="_blank"
            >
              wit3
            </a>
          </td>
        </div>
        <div
          class="responsive-table"
          data-v-a63dd56f=""
        >
          <th
            class="label"
            data-v-a63dd56f=""
          >
            Amount
          </th>
          <td
            class="px-md py-md [&&]:sm:pt-sm whitespace-nowrap"
            data-v-a63dd56f=""
          >
            <span
              data-v-a63dd56f=""
            >
              1
            </span>
          </td>
        </div>
        
        
        <td
          class="sm:hidden px-md py-md [&&]:sm:pt-sm whitespace-nowrap last:text-end"
          data-v-a63dd56f=""
        >
          <span
            data-v-a63dd56f=""
          >
            1
          </span>
        </td>
        <td
          class="sm:hidden px-md py-md [&&]:sm:pt-sm whitespace-nowrap last:text-end"
          data-v-a63dd56f=""
        >
          <a
            class="text-wit-blue-600 dark:text-wit-blue-500"
            data-v-a63dd56f=""
            href="https://"
            target="_blank"
          >
            wit3
          </a>
        </td>
        <td
          class="sm:hidden px-md py-md [&&]:sm:pt-sm whitespace-nowrap last:text-end"
          data-v-a63dd56f=""
        >
          <span
            data-v-a63dd56f=""
          >
            1
          </span>
        </td>
        
      </tr>
      <tr
        class="transition-all duration-200 even:bg-white-200 dark:even:bg-black-600 sm:grid sm:grid-col-1 sm:py-md"
        data-v-a63dd56f=""
      >
        
        <div
          class="responsive-table"
          data-v-a63dd56f=""
        >
          <th
            class="label"
            data-v-a63dd56f=""
          >
            Rank
          </th>
          <td
            class="px-md py-md [&&]:sm:pt-sm whitespace-nowrap"
            data-v-a63dd56f=""
          >
            <span
              data-v-a63dd56f=""
            >
              2
            </span>
          </td>
        </div>
        <div
          class="responsive-table"
          data-v-a63dd56f=""
        >
          <th
            class="label"
            data-v-a63dd56f=""
          >
            Address
          </th>
          <td
            class="px-md py-md [&&]:sm:pt-sm whitespace-nowrap"
            data-v-a63dd56f=""
          >
            <a
              class="text-wit-blue-600 dark:text-wit-blue-500"
              data-v-a63dd56f=""
              href="https://"
              target="_blank"
            >
              wit1
            </a>
          </td>
        </div>
        <div
          class="responsive-table"
          data-v-a63dd56f=""
        >
          <th
            class="label"
            data-v-a63dd56f=""
          >
            Amount
          </th>
          <td
            class="px-md py-md [&&]:sm:pt-sm whitespace-nowrap"
            data-v-a63dd56f=""
          >
            <span
              data-v-a63dd56f=""
            >
              2
            </span>
          </td>
        </div>
        
        
        <td
          class="sm:hidden px-md py-md [&&]:sm:pt-sm whitespace-nowrap last:text-end"
          data-v-a63dd56f=""
        >
          <span
            data-v-a63dd56f=""
          >
            2
          </span>
        </td>
        <td
          class="sm:hidden px-md py-md [&&]:sm:pt-sm whitespace-nowrap last:text-end"
          data-v-a63dd56f=""
        >
          <a
            class="text-wit-blue-600 dark:text-wit-blue-500"
            data-v-a63dd56f=""
            href="https://"
            target="_blank"
          >
            wit1
          </a>
        </td>
        <td
          class="sm:hidden px-md py-md [&&]:sm:pt-sm whitespace-nowrap last:text-end"
          data-v-a63dd56f=""
        >
          <span
            data-v-a63dd56f=""
          >
            2
          </span>
        </td>
        
      </tr>
      <tr
        class="transition-all duration-200 even:bg-white-200 dark:even:bg-black-600 sm:grid sm:grid-col-1 sm:py-md"
        data-v-a63dd56f=""
      >
        
        <div
          class="responsive-table"
          data-v-a63dd56f=""
        >
          <th
            class="label"
            data-v-a63dd56f=""
          >
            Rank
          </th>
          <td
            class="px-md py-md [&&]:sm:pt-sm whitespace-nowrap"
            data-v-a63dd56f=""
          >
            <span
              data-v-a63dd56f=""
            >
              3
            </span>
          </td>
        </div>
        <div
          class="responsive-table"
          data-v-a63dd56f=""
        >
          <th
            class="label"
            data-v-a63dd56f=""
          >
            Address
          </th>
          <td
            class="px-md py-md [&&]:sm:pt-sm whitespace-nowrap"
            data-v-a63dd56f=""
          >
            <a
              class="text-wit-blue-600 dark:text-wit-blue-500"
              data-v-a63dd56f=""
              href="https://"
              target="_blank"
            >
              wit2
            </a>
          </td>
        </div>
        <div
          class="responsive-table"
          data-v-a63dd56f=""
        >
          <th
            class="label"
            data-v-a63dd56f=""
          >
            Amount
          </th>
          <td
            class="px-md py-md [&&]:sm:pt-sm whitespace-nowrap"
            data-v-a63dd56f=""
          >
            <span
              data-v-a63dd56f=""
            >
              3
            </span>
          </td>
        </div>
        
        
        <td
          class="sm:hidden px-md py-md [&&]:sm:pt-sm whitespace-nowrap last:text-end"
          data-v-a63dd56f=""
        >
          <span
            data-v-a63dd56f=""
          >
            3
          </span>
        </td>
        <td
          class="sm:hidden px-md py-md [&&]:sm:pt-sm whitespace-nowrap last:text-end"
          data-v-a63dd56f=""
        >
          <a
            class="text-wit-blue-600 dark:text-wit-blue-500"
            data-v-a63dd56f=""
            href="https://"
            target="_blank"
          >
            wit2
          </a>
        </td>
        <td
          class="sm:hidden px-md py-md [&&]:sm:pt-sm whitespace-nowrap last:text-end"
          data-v-a63dd56f=""
        >
          <span
            data-v-a63dd56f=""
          >
            3
          </span>
        </td>
        
      </tr>
      
    </tbody>
  </table>
  <!--v-if-->
  
</div>
`;
